### ISSUE TITLE
text boxes

### ISSUE THREAD

**Issue by mujihina** [Windower/Lua#811]
Something changed recently.....
- With scaling > 1.0, when dragging text objects, one has to click much more to the left/top than before.  (The farther you drag, the more the offset grows)
- Regardless of scaling, half the time now, clicking on the text object generates this runtime error:
  lua:462: attempt to compare nil with number.

**Comment by alphaONE2** [comment]
Please limit each ticket to a single issue; combining multiple issues like this makes managing the issue tracker much more complicated.

The first issue is because mouse coordinates are not scaled with the UI. We should probably scale the coordinates in LuaCore so that Addons don't need to deal with the extra complexity.

I'm not sure what's causing the second issue.

**Comment by z16** [comment]
Yeah, I'll add the first thing to LuaCore today.

Does the latter issue also only appear with higher scaling? I'm pretty sure it's referring to `Windower/addons/libs/texts.lua`, line `462`:

``` lua
            and (pos_x <= x and x <= pos_x + off_x
```

So it's also coordinate related. It uses `windower.texts.get_extents` and `windower.text.get_location`, so maybe one of them is returning `nil`. Can you try adding this line before that and see what it prints out:

``` lua
            print(pos_x, pos_y, off_x, off_y, x, y)
```

**Comment by mujihina** [comment]
text boxes don't even show today ;)

**Comment by z16** [comment]
I didn't change LuaCore yet at all, I was gonna ask you to try [this one](https://googledrive.com/host/0B1-GYvFFRpK6WlVQZnBKbzdycTA/LuaCore.dll) and let me know if that helps. Do you mean text boxes don't show up at all for you right now?

**Comment by mujihina** [comment]
Yes.  None of my addons can show text objects on screen.

output of the pos_x, pos_y, off_x, off_y, x, y of a text object shows as:

pos_x: function 23F35798  (shouldn't this be a number? I set it to 100 during creation)
pos_x: function 23F35178 (shouldn't this be a number? I set it to 100 during creation)
off_x: nil
off_y: nil
x: nil
y: nil

I was going to switch to the non-dev version of windower to get the output...but it seems Norton removed windower.exe because of Suspicious.Cloud.9. I'll submit a new issue for that.

**Comment by Byrth** [comment]
Don't bother submitting anything for Norton. We can't control what they do.

**Comment by z16** [comment]
Yes, they should be numbers, but this doesn't happen for me... I can still see text boxes just fine. Try just entering this in eval:

``` lua
texts.new('Sample text'):show()
```

This should create a visible text box at the top-left border of the screen.

Are you still using UI scale larger than `1`? If so, can you see if it happens with UI scale set to `1`? Are you using the LuaCore I gave you? If so, can you reset to the regular LuaCore? Maybe that is causing issues, since I added some code to X/Y coordinate handling. To reset just delete `LuaCore.dll` and restart `Windower.exe` (or rename `LuaCore.dll` if you have a Windower instance running and then restart `Windower.exe`).

**Comment by mujihina** [comment]
hmm... eval is loaded and everything, but
texts.new('Sample text'):show()
doesn't show anything

I also removed LuaCore.dll and Hook.dll (you gave me a hook.dll a while back) and restarted windower.exe

Same behavior. No text boxes from eval or my existing addons.

However, I noticed that packetviewer STILL shows text boxes. After a quick investigation, it seems that without root settings, it won't work anymore.

e.g.:
something = texts.new (whatever, settings.something) -- It no longer works.
but
something = texts.new (whatever, settings.something, settings) -- This works.

The last change to texts.lua seem relevant.
When I add the recently removed line:
apply_settings(t, t._settings)

my addons work again, and I can also create and show text objects with eval.
